lab_3/main.py

```diff
@@ -17,39 +19,66 @@ def tokenize_by_sentence(text: str) -> tuple:
          (('_', 'h', 'e', '_'), ('_', 'i', 's', '_'), ('_', 'h', 'a', 'p', 'p', 'y', '_'))
          )
     """
-    pass
+    if not isinstance(text, str) or text == '':
+        return ()
+
+    text_outputt = re.split('[.] ', text)
+    token_text = []
+    for i in text_outputt:
+        w_list = re.sub('[^a-z \n]', '', i.lower()).split()
+        if w_list == []:
+            return ()
+        token_text.append(tuple(tuple(['_'] + list(w) + ['_']) for w in w_list))
+
+    return tuple(token_text)
 
 
 # 4
 class LetterStorage:
 
     def __init__(self):
-        pass
+        self.storage = {}
 
     def _put_letter(self, letter: str) -> int:
         """
         Puts a letter into storage, assigns a unique id
         :param letter: a letter
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        len_storage = len(self.storage)
+        if not isinstance(letter, str) or letter == '':
+            return 1
+        if letter not in self.storage:
+            self.storage[letter] = len_storage + 1
+            return 0
+        return 0
 
     def get_id_by_letter(self, letter: str) -> int:
         """
         Gets a unique id by a letter
         :param letter: a letter
         :return: an id
         """
-        pass
+        if letter in self.storage:
+            letter_id = self.storage.get(letter)
+            return letter_id
+        return -1
 
     def update(self, corpus: tuple) -> int:
         """
         Fills a storage by letters from the corpus
         :param corpus: a tuple of sentences
         :return: 0 if succeeds, 1 if not
         """
-        pass
+        if isinstance(corpus, tuple):
+            for w in corpus:
```

Comment:
a corpus consists of sentences, sentences consist of words, words consist of letters